spring-cloud-gcp-kms/src/main/java/com/google/cloud/spring/kms/KmsTemplate.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.kms;
+
+import java.util.Base64;
+
+import com.google.cloud.kms.v1.CryptoKeyName;
+import com.google.cloud.kms.v1.DecryptRequest;
+import com.google.cloud.kms.v1.DecryptResponse;
+import com.google.cloud.kms.v1.EncryptRequest;
+import com.google.cloud.kms.v1.EncryptResponse;
+import com.google.cloud.kms.v1.KeyManagementServiceClient;
+import com.google.cloud.spring.core.GcpProjectIdProvider;
+import com.google.common.hash.Hashing;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Int64Value;
+
+/**
+ * Offers convenience methods for performing common operations on KMS including
+ * encrypting and decrypting text.
+ *
+ * @author Emmanouil Gkatziouras
+ */
+public class KmsTemplate implements KmsOperations {
+
+	private final KeyManagementServiceClient client;
+
+	private final GcpProjectIdProvider projectIdProvider;
+
+	public KmsTemplate(
+			KeyManagementServiceClient keyManagementServiceClient,
+			GcpProjectIdProvider projectIdProvider) {
+		this.client = keyManagementServiceClient;
+		this.projectIdProvider = projectIdProvider;
+	}
+
+	public String encrypt(String cryptoKey, String plaintext) {
+		CryptoKeyName cryptoKeyName = KmsPropertyUtils.getCryptoKeyName(cryptoKey, projectIdProvider);
+
+		ByteString plaintextByteString = ByteString.copyFromUtf8(plaintext);
+
+		long crc32c = longCrc32c(plaintextByteString);
+
+		EncryptRequest request = EncryptRequest.newBuilder()
+				.setName(cryptoKeyName.toString())
+				.setPlaintext(plaintextByteString)
+				.setPlaintextCrc32C(
+						Int64Value.newBuilder().setValue(crc32c).build())
+				.build();
+
+		EncryptResponse response = client.encrypt(request);
+		assertCrcMatch(response);
+
+		return encodeBase64(response);
```

Comment:
Is the Base64 encoding and decoding really necessary?
It adds overhead that some users may not like.
I think there are two main use cases: (1) encrypting/decoding utf8-encoded text, and (2) encrypting/decrypting binary data (byte[]).

So, I would propose just having 4 methods:
```java
byte[] encryptBytes(String cryptoKey, byte[] plaintext);
byte[] encryptText(String cryptoKey, String plaintext); // UTF-8 encoded text
byte[] decryptBytes(String cryptoKey, byte[] ciphertext);
String decryptText(String cryptoKey, byte[] ciphertext); // returns UTF-8 encoded text
```

What do you think?

Reply:
Understood. Will implement those methods however when it comes to the problem described on [previous](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/175#discussion_r542753418) comment, encoding to UTF-8 encode I fail to decrypt the UTF-8 encoded text.

```
		KeyManagementServiceClient client = KeyManagementServiceClient.create();

		CryptoKeyName cryptoKeyName = CryptoKeyName.newBuilder()
				.setProject("project")
				.setLocation("europe-west2")
				.setKeyRing("spring-cloud-gcp")
				.setCryptoKey("key-id")
				.build();

		String encrypted = client.encrypt(cryptoKeyName.toString(), ByteString.copyFromUtf8( "1234")).getCiphertext().toStringUtf8();
		String decrypted = client.decrypt(cryptoKeyName.toString(), ByteString.copyFromUtf8(encrypted)).getPlaintext().toStringUtf8();
```

The decryption of the String that was UTF-8 after it's encryption will fail. I am not sure what the sdk does behind the scenes but that's why I resided on base64.

Would it make sense to just have the byte methods and then up to the user on how to proceed storing and using? 
```
byte[] encryptBytes(String cryptoKey, byte[] plaintext);
byte[] decryptBytes(String cryptoKey, byte[] ciphertext);
```

Additionally those method can be put for an extra for convenience 

```
byte[] encryptText(String cryptoKey, String plaintext); // Base64 encoded text
String decryptText(String cryptoKey, byte[] ciphertext); // returns Base64
```

Reply:
Let me play around with this.

Reply:
I think there is an issue with converting between `byte[]` and `String` back and forth.
The ciphertext should probably be kept as `byte[]` because it's binary.
This works:
```java
String plaintext = "1234";
byte[] encrypted = client.encrypt(cryptoKeyName, ByteString.copyFromUtf8(plaintext)).getCiphertext().toByteArray();
String decrypted = client.decrypt(cryptoKeyName, ByteString.copyFrom(encrypted)).getPlaintext().toStringUtf8();
assertThat(plaintext).isEqualTo(decrypted);
```

Reply:
Yeap it seems the `.toStringUtf8()` corrupts the result (that's why base64 works, just carries the bytes as is).

Going for this one as proposed, since it will work and sticks to bytes only.

```
byte[] encryptBytes(String cryptoKey, byte[] plaintext);
byte[] decryptBytes(String cryptoKey, byte[] ciphertext);
byte[] encryptText(String cryptoKey, String plaintext); // UTF-8 encoded text
String decryptText(String cryptoKey, byte[] ciphertext); // returns UTF-8 encoded text
```

Reply:
Exactly, UTF-8 encoding is not always reversible, although it is for Base64.
For example, this test fails:
```java
byte[] bytes = new byte[] {10, -10};
String stringFromBytes = new String(bytes);
assertThat(bytes).isEqualTo(stringFromBytes.getBytes());
```

Reply:
The methods have been updated as discussed. The implementation is bas64 clear.
The sample in order to make a copy paste feasible uses bas64 encoding on the controller.